src/main/java/gg/solarmc/kitpvp/handler/KillHandler.java

```diff
@@ -0,0 +1,204 @@
+
+        String victimName = victim.getName();
+        // Playing a little bit loose, but still correct, with thread safety
+        List<Runnable> postTransactCallbacks = new ArrayList<>();
+        return dataCenter.runTransact((tx) -> {
+            KitPvp killerData = killer.getSolarPlayer().getData(KitPvpKey.INSTANCE);
+            KitPvp victimData = victim.getSolarPlayer().getData(KitPvpKey.INSTANCE);
+
+            /*
+             * Checklist:
+             * Update killstreak for killer
+             * Update experience for killer
+             * Reset killstreak for victim
+             * Increment killer kills and victim deaths
+             * Reward each assistant and increment their assists
+             */
+
+            // Update killer killstreak
+            int killerNewKillstreak = killerData.addKillstreaks(tx, 1).newValue();
+            postTransactCallbacks.add(newKillstreakReached(killer, killerNewKillstreak));
+
+            // Update killer experience
+            {
+                StatisticResult addExperienceResult = killerData.addExperience(tx, violence().experiencePerKill());
+                int killerNewLevel = levelCalculator.calculateLevel(addExperienceResult.newValue());
+                int killerOldLevel = levelCalculator.calculateLevel(addExperienceResult.oldValue());
+                if (killerNewLevel != killerOldLevel) {
+                    postTransactCallbacks.add(levelUp(killer, killerOldLevel, killerNewLevel));
+                }
+            }
+
+            // Reset victim killstreak and send message
+            {
+                int victimLostKillstreak = victimData.resetCurrentKillstreaks(tx);
+                postTransactCallbacks.add(previousKillstreakLost(victim, victimLostKillstreak));
+            }
+
+            // Update statistics and add assist rewards
+            {
+                killerData.addKills(tx, 1);
+                victimData.addDeaths(tx, 1);
+                BigDecimal rewardPerAssist = violence().killRewards().assistReward();
+                for (Player assistant : assistants) {
+                    assistant.getSolarPlayer().getData(KitPvpKey.INSTANCE).addAssists(tx, 1);
+                    DepositResult depositResult = assistant.getSolarPlayer().getData(CreditsKey.INSTANCE)
+                            .depositBalance(tx, rewardPerAssist);
+                    postTransactCallbacks.add(assistedKillReward(assistant, victimName, rewardPerAssist, depositResult));
+                }
+            }
+
+            BigDecimal killReward = violence().killRewards().killReward().findValue(killerNewKillstreak);
+            DepositResult depositResult = killer.getSolarPlayer().getData(CreditsKey.INSTANCE)
+                    .depositBalance(tx, killReward);
+            postTransactCallbacks.add(mainKillReward(killer, victimName, killReward, depositResult));
+
+        }).thenRun(() -> {
+            for (Runnable callback : postTransactCallbacks) {
```

Comment:
Didnt think of this

Reply:
Preserves code readability reasonably though - why is this in thenRun though

Reply:
shouldn't be thenRunSync?

Reply:
Thanks to https://github.com/SolarMC-Dev/SolarPaper-API/commit/fb2d80601e5b164f97f79b6924c4f777c447a849 sendMessage with Components is guaranteed to be thread-safe (and indeed the implementation is)